src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs

```diff
@@ -272,26 +287,56 @@ private void SolutionManager_CacheUpdated(object sender, NuGetEventArgs<string>
                     // We also refresh the UI if projectFullPath is not present.
                     if (!projectContainsFullPath || projectFullName == eventProjectFullName)
                     {
-                        RefreshWhenNotExecutingAction();
+                        RefreshWhenNotExecutingAction(RefreshOperationSource.CacheUpdated);
+                    }
+                    else
+                    {
+                        EmitRefreshEvent(timeSpan, RefreshOperationSource.CacheUpdated, RefreshOperationStatus.NotApplicable);
                     }
                 }
             }
+            else
+            {
+                EmitRefreshEvent(timeSpan, RefreshOperationSource.CacheUpdated, RefreshOperationStatus.NoOp);
+            }
         }
 
-        private void RefreshWhenNotExecutingAction()
+        private void RefreshWhenNotExecutingAction(RefreshOperationSource source)
         {
+            var timeSpan = GetTimeSinceLastRefreshAndRestart();
             // Only refresh if there is no executing action. Tell the operation execution to refresh when done otherwise.
             if (!_isExecutingAction)
             {
                 Refresh();
+                EmitRefreshEvent(timeSpan, source, RefreshOperationStatus.Success);
             }
             else
             {
                 _isRefreshRequired = true;
+                EmitRefreshEvent(timeSpan, source, RefreshOperationStatus.NoOp);
+
             }
         }
 
-        public PackageRestoreBar RestoreBar => _restoreBar;
+        private void EmitRefreshEvent(TimeSpan timeSpan, RefreshOperationSource refreshOperationSource, RefreshOperationStatus status)
+        {
+            TelemetryActivity.EmitTelemetryEvent(
+                                new PackageManagerUIRefreshEvent(
+                                    _sessionGuid,
+                                    Model.IsSolution,
+                                    refreshOperationSource,
+                                    status,
+                                    _topPanel.Filter.ToString(),
+                                    timeSpan
+                                    ));
+        }
+
+        private TimeSpan GetTimeSinceLastRefreshAndRestart()
+        {
+            var elapsed = _lastRefresh.Elapsed;
+            _lastRefresh.Restart();
```

Comment:
Are operations on `_lastRefresh` thread safe?  Typically, instance members are not thread safe.  Concurrent operations on `_lastRefresh` might result in some unexpected math.

Reply:
We discussed some offline, but thinking about it some more, we could always get some discrepancy and weird math in some situations. 

The root cause is that there's a gap between when we measure and when we make the decision about a refresh occuring. 
In the sandwich scenario for 2 threads that are taking the time and deciding whether to run a refresh or not, this will always lead to issues. 

In fact I'm not sure there's ever a right way to time this. Even if we move the time taking right after we decide whether a request is gonna trigger, there's still a potential for inconsistency, and there's no guarantee that these decisions whether to refresh or not are related in any way. 

I think we'll just have to take these numbers with a grain of salt, understanding it's limitations. 

Happy to discuss in person.